blog/2022-11-03-short-circuiting.md

```diff
@@ -0,0 +1,156 @@
+---
+title: Array short-circuiting
+author: Philip Munksgaard
+description: In-place updates without overhead
+---
+
+# Situation
+
+Futhark is a parallel high-level array-oriented language.  Parallelism is explicitly
+expressed by the user in terms of the second-order array combinators such as
+`map`, `reduce` and `scan`.  Because parallelism is explicitly described, the
+compiler is able to make certain guarantees about the parallelism, such that the
+guarantee that there are no data races in the code.
+
+To see why this matters, consider this simple stencil function:
+
+```
+def stencil [n] (xs: *[n]i64): [n]i64 =
+  xs with [1:n-1] = map (\i -> xs[i-1] + xs[i] + xs[i+1]) (1..<n-1)
+```
+
+In a language without any data-race guarantees, or we were to implement this
+function as a parallel for loop in an imperative language, we would have
+data-dependency errors.  Specifically, it is possible for one thread to update
```

Comment:
I'd suggest a remark that in other languages you might implement this in-place, which is dangerous, but in a functional language *nothing* is in-place, which is on the other hand sometimes too conservative and slow.